### PR TITLE
ENH: Provide more control over elemental block indexing.

### PR DESCRIPTION
**Pull request by GuySartorelli**
Provides a new configuration variable to exclude specific elemental block classes from being indexed in search.
Also provides a new extension point for modifying exactly what gets indexed for each block.

### CODE REVIEW

File: src/Extensions/ElementalPageExtension.php [hunk 1]
```diff
@@ -43,7 +49,16 @@ public function getElementsForSearch()
                 /** @var ElementalArea $area */
                 $area = $this->owner->$key();
                 if ($area) {
-                    $output[] = strip_tags($area->forTemplate());
+                    // Only render the elemental blocks which are not excluded from search and don't pull from a page.
+                    foreach ($area->Elements() as $element) {
+                        if (!in_array($element->ClassName, Config::inst()->get(static::class, 'elements_excluded_from_search'))) {
+                            $forIndexing = $element->forTemplate();
```
Comment (michalkleiner) [inline src/Extensions/ElementalPageExtension.php:55]: This always renders the element and then it decides whether it will throw it away based on the extension point.
Could it be reworked so excluded elements don't even render in the first place?

I can imagine something like adding `public function getIsIndexable(): bool` to the BaseElement and then each element can provide custom logic that would be used here to skip the rendering.

Comment (GuySartorelli) [reply src/Extensions/ElementalPageExtension.php:55]: >This always renders the element and then it decides whether it will throw it away based on the extension point.

The intention was that it excludes it if it's in the `elements_excluded_from_search` array, but then on a block-by-block basis, developers have even more control. So it might be a block which isn't excluded by class, but perhaps there is a rule where if some fields are not filled in it shouldn't be displayed (or indexed). So in that case the developer might choose to not index that specific block.
They can also choose to just modify what is indexed, in case the display of the block is not the same as what should be indexed for the block (I have at least one block where some fields are pulled from other sources, so I wouldn't want to index those fields as a part of this block).
If on the other hand the block's class is in the `elements_excluded_from_search` array, that if statement will return false and it will automatically be excluded.

I like the idea of a `getIsIndexable` method in principle provided it has an extension point in it to allow modifying that value (I use a lot of third-party blocks and want to define what is indexable for my project). But I think being able to modify _what_ is indexed per block is also important.

Comment (michalkleiner) [reply src/Extensions/ElementalPageExtension.php:55]: Fair point about the modification, but basically at that point you're dealing with rendered markup, not structured data. So if you wanted to avoid e.g. an API call, that would be too late anyway with the above code.

There's an epic somewhere to provide a better content API that would allow for that granularity.

**Comment by GuySartorelli** [comment]
Not sure when I'll get a chance to update this to be in line with @michalkleiner's recommendations - setting to draft until then.

File: src/Extensions/ElementalPageExtension.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Extensions/ElementalPageExtension.php]: I've made changes to allow per-block configuration both for whether a block is indexable in the first place, and if so, what content exactly will be indexed.
I think this is a much more sensible way of handling it. Please review.

**Comment by GuySartorelli** [comment]
Tests are failing, but I don't see how the changes I've made could have caused that.... can someone who understands the `ElementSiteTreeFilterSearchTest` please take a look?

File: src/Models/BaseElement.php [hunk 2]
```diff
@@ -457,6 +466,33 @@ public function Top()
...
+     * Provides content to be indexed in search.
+     *
+     * @return string|null HTML
+     */
+    public function forSearchIndex()
+    {
+        // Strips tags but be sure there's a space between words.
+        $index = trim(strip_tags(str_replace('<', ' <', $this->forTemplate())));
+        // Allow projects to update indexable content of third-party elements.
+        $this->extend('updateSearchIndex', $index);
```
Comment (michalkleiner) [inline src/Models/BaseElement.php:492]:
```suggestion
        $this->extend('updateContentForSearchIndex', $content);
```

File: src/Models/BaseElement.php [hunk 3]
```diff
@@ -457,6 +466,33 @@ public function Top()
...
+    {
+        return (bool)$this->config()->get('is_indexable');
+    }
+
+    /**
+     * Provides content to be indexed in search.
+     *
+     * @return string|null HTML
+     */
+    public function forSearchIndex()
```
Comment (michalkleiner) [inline src/Models/BaseElement.php:487]:
```suggestion
    public function getContentForSearchIndex()
```

File: src/Models/BaseElement.php [hunk 4]
```diff
@@ -457,6 +466,33 @@ public function Top()
...
+     * Determines whether this elemental block is indexable in search.
+     *
+     * By default, this uses the configurable variable is_indexable, but
+     * this method can be overridden to provide more complex logic if required.
+     *
+     * @return boolean
+     */
+    public function getIsIndexable(): bool
+    {
+        return (bool)$this->config()->get('is_indexable');
```
Comment (michalkleiner) [inline src/Models/BaseElement.php:479]: Should this use the uninherited config so that inheritance wouldn't play role in excluding certain class from indexing?

Comment (GuySartorelli) [reply src/Models/BaseElement.php:479]: Ooh yes, that's a good idea. I'll make that adjustment.

File: src/Extensions/ElementalPageExtension.php [hunk 5]
```diff
@@ -52,7 +59,7 @@ public function getElementsForSearch()
             // CMS layout can break on the response. (SilverStripe 4.1.1)
             SSViewer::set_themes($oldThemes);
         }
-        return implode($output);
+        return implode(' ... ', $output);
```
Comment (michalkleiner) [inline src/Extensions/ElementalPageExtension.php:62]: This will probably only need to be a space, we can't be adding dots into the content (not even an ellipsis).

Comment (GuySartorelli) [reply src/Extensions/ElementalPageExtension.php:62]: Only adding a single space will help to ensure words aren't immediately adjacent to one another, but it will also mean that sometimes searching for a phrase could result in incorrectly finding a phrase between the end of one block and the start of another...

Would you object to me making this configurable? I will then have it as a space by default, but developers can choose to use whatever delimiter they choose between blocks.

Comment (michalkleiner) [reply src/Extensions/ElementalPageExtension.php:62]: Yep, get your point around phrase queries. This mechanism of concatenating block content isn't ideal, same as the rendered version. The best way would be to where the content is added into multi-value solr content field separately, not as a single value.
Configurable delimiter will do for now.

File: src/Models/BaseElement.php [hunk 6]
```diff
@@ -169,6 +170,15 @@ class BaseElement extends DataObject
...
     private static $displays_title_in_template = true;
 
+    /**
+     * Determines whether a block should be indexable in search.
+     *
+     * @config
+     * @var boolean
+     * @see ElementalPageExtension::getElementsForSearch()
+     */
+    private static $is_indexable = true;
```
Comment (emteknetnz) [inline src/Models/BaseElement.php:180]:
```suggestion
    private static $search_indexable = true;
```
We probably want the word 'search' explicitly in the var name, so it's not confused with a database index.  Can also remove the `is_` prefix

File: src/Extensions/ElementalPageExtension.php [hunk 7]
```diff
@@ -25,6 +25,17 @@ class ElementalPageExtension extends ElementalAreasExtension
...
+    /**
+     * The delimiter to separate distinct elements in indexed content.
+     *
+     * When using the getElementsForSearch() method to index all elements in a single field,
+     * a custom delimiter can be used help to avoid false positive results for phrase queries.
+     *
+     * @config
+     * @var string
+     */
+    private static $delimiter_for_indexed_elements = ' ';
```
Comment (emteknetnz) [inline src/Extensions/ElementalPageExtension.php:37]: What's a use case for this to not be a space?

I can't really think of what we'd want to do something like put a dash `-` between element content in search results?

If we do retain this configurable var, could we rename it to `$search_index_element_delimiter`?

File: src/Models/BaseElement.php [hunk 8]
```diff
@@ -470,6 +480,33 @@ public function Top()
...
 
+    /**
+     * Determines whether this elemental block is indexable in search.
+     *
+     * By default, this uses the configurable variable is_indexable, but
+     * this method can be overridden to provide more complex logic if required.
+     *
+     * @return boolean
+     */
+    public function getIsIndexable(): bool
```
Comment (emteknetnz) [inline src/Models/BaseElement.php:491]:
```suggestion
    public function getSearchIndexable(): bool
```

File: src/Extensions/ElementalPageExtension.php [hunk 9]
```diff
@@ -43,7 +54,14 @@ public function getElementsForSearch()
                 /** @var ElementalArea $area */
                 $area = $this->owner->$key();
                 if ($area) {
-                    $output[] = strip_tags($area->forTemplate());
+                    foreach ($area->Elements() as $element) {
+                        if ($element->getIsIndexable()) {
+                            $forIndexing = $element->getContentForSearchIndex();
```
Comment (emteknetnz) [inline src/Extensions/ElementalPageExtension.php:59]:
```suggestion
                            $content = $element->getContentForSearchIndex();
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Seems good, some requested changes

Could you also add some unit tests?  Particularly for BaseElement::getContentForSearchIndex().

Let me know if not sure how to go about setting up tests for this and I'll point you in the right direction

File: src/Models/BaseElement.php [hunk 10]
```diff
@@ -470,6 +480,33 @@ public function Top()
...
+    {
+        return (bool)$this->config()->get('is_indexable', Config::UNINHERITED);
+    }
+
+    /**
+     * Provides content to be indexed in search.
+     *
+     * @return string|null HTML
+     */
+    public function getContentForSearchIndex()
```
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply src/Models/BaseElement.php:501]: Would you be okay with `?string`? See discussion in PHPDoc conversation.

File: src/Models/BaseElement.php [hunk 11]
```diff
@@ -470,6 +480,33 @@ public function Top()
...
+     */
+    public function getIsIndexable(): bool
+    {
+        return (bool)$this->config()->get('is_indexable', Config::UNINHERITED);
+    }
+
+    /**
+     * Provides content to be indexed in search.
+     *
+     * @return string|null HTML
```
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply src/Models/BaseElement.php:499]: This is intentionally nullable so that developers can return null from an overridden method in a subclass if they explicitly don't want content to be indexed. Let me know if you're okay with that approach - I should probably specify that in the PHPDoc for clarity.

File: src/Extensions/ElementalPageExtension.php (hunk 7 above)
Comment (GuySartorelli) [reply src/Extensions/ElementalPageExtension.php:37]: I initially had it as ellipses but @michalkleiner rightly pointed out that may be unexpected for some developers. In my projects I will want to use an ellipses so that search results show a clear separation between content from separate elemental blocks.

**Comment by GuySartorelli** [comment]
@emteknetnz I'll make the requested changes as time permits, hopefully in the next week. Please refer to the discussion on the PHPDoc and let me know on that one.

I note there is also a conflict now so I'll resolve that at the same time.

File: src/Models/BaseElement.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Models/BaseElement.php:499]: What's the functional difference between `null` and `''` (empty string) though?  Is the something in the FullTextSearch module (or another search module?) that explicitly tests for `null`?

File: src/Extensions/ElementalPageExtension.php (hunk 7 above)
Comment (emteknetnz) [reply src/Extensions/ElementalPageExtension.php:37]: OK the ellipses use case makes sense.  Yes rename thanks

File: src/Models/BaseElement.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply src/Models/BaseElement.php:499]: It's a semantic difference rather than a functional one so I'm not too worried about it at the end of the day.

[the remark this answers is not in the document]
Comment (GuySartorelli) [reply src/Models/BaseElement.php:499]: I've left this nullable for now and made the typehint nullable because it is quite common for developers to return null from methods such as this one in ss projects - happy to remove nullability in the end if you want, but I had to make a call.

**Comment by GuySartorelli** [comment]
@emteknetnz I've made the requested changes (though see [discussion about phpdoc return type](#discussion_r788320014) which affects the [related typehint](#discussion_r788319930) except tests. I've also added some documentation.

> Let me know if not sure how to go about setting up tests for this and I'll point you in the right direction
Yes please. I thought the tests would be relatively straight forward but I spent a good 15 minutes and then realised I'm not really sure what exactly I _want_ to be testing... if you could give me a direction (even just some method signatures or comments to build off) that would be great.

**Comment by GuySartorelli** [comment]
Failed tests are because I forgot `Extension` isn't inherently `Configurable` - I'll fix that at the same time as writing tests once I know what tests to write.

File: src/Models/BaseElement.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (emteknetnz) [reply src/Models/BaseElement.php]: Returning a blank string also won't get indexed :-) i.e. what search term could and end user put into a search box to match blank?  I just can't see a use-case where `null` does something that `''` doesn't, so I'd prefer to keep the API simple and just have a `string` return type

File: src/Models/BaseElement.php [hunk 12]
```diff
@@ -476,6 +486,33 @@ public function Top()
...
+     * Determines whether this elemental block is indexable in search.
+     *
+     * By default, this uses the configurable variable search_indexable, but
+     * this method can be overridden to provide more complex logic if required.
+     *
+     * @return boolean
+     */
+    public function getSearchIndexable(): bool
+    {
+        return (bool)$this->config()->get('search_indexable', Config::UNINHERITED);
```
Comment (emteknetnz) [inline src/Models/BaseElement.php:499]:
```suggestion
        return (bool) $this->config()->get('search_indexable', Config::UNINHERITED);
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Tiny linting thing and change return type to `string`.  Cheers

**Comment by emteknetnz** [comment]
@GuySartorelli This would be a good level of test coverage for the work you've done here:

Setup:
- Create a couple of test elements of ElementContent (comes with `HTML`  db field).
- Create a test Extension and put it on ElementContent that implements the `updateContentForSearchIndex` hook.
- Add elements to an ElementalArea + SiteTree

See and example [here](https://github.com/silverstripe/silverstripe-asset-admin/blob/1/tests/php/Forms/FileFormBuilderTest.php#L121) for how to add test extensions in a unit test

To change the config for the delimiter in the unit test, simply `MyClass::config()->set('config_var', $value);` - it will reset between tests so won't affect other tests (we'll I'm at least 99% sure on that :-)

Tests:
a) Test that changing `search_index_element_delimiter` delimits the output
b) Test that getContentForSearchIndex adds spaces to HTML
c) Test that an extension on getContentForSearchIndex updates the content

File: src/Models/BaseElement.php (hunk 12 above)
Comment (GuySartorelli) [reply src/Models/BaseElement.php:499]: Darnit I felt like I had missed something 😅

File: src/Models/BaseElement.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply src/Models/BaseElement.php]: I can agree to that logic! I'll make the change.

**Comment by GuySartorelli** [comment]
Tests failed - working on that now.

**Comment by GuySartorelli** [comment]
@emteknetnz I've added the requested tests and those are passing. There is one test that's failing, but I haven't touched that test and (from what I can tell) none of the changes I've made should be causing this to fail. Can you please take a look?

**Comment by emteknetnz** [comment]
@GuySartorelli found the issue for the travis failure, it was an existing bug - have created a PR https://github.com/silverstripe/silverstripe-elemental/pull/957

**Review by emteknetnz — APPROVED** [review]
Travis failure is an existing bug that this PR highlighted, separate PR to fix it is here https://github.com/silverstripe/silverstripe-elemental/pull/957
